pylint/checkers/variables.py

```diff
@@ -3401,5 +2642,754 @@ def visit_const(self, node: nodes.Const) -> None:
+                and frame is defframe
+                and defframe.parent_of(node)
+                and stmt is defstmt
+            ):
+                # Single statement if, with assignment expression on same
+                # line as assignment
+                # x = b if (b := True) else False
+                maybe_before_assign = False
+            elif (
+                isinstance(  # pylint: disable=too-many-boolean-expressions
+                    defnode, nodes.NamedExpr
+                )
+                and frame is defframe
+                and defframe.parent_of(stmt)
+                and stmt is defstmt
+                and (
+                    (
+                        defnode.lineno == node.lineno
+                        and defnode.col_offset < node.col_offset
+                    )
+                    or (defnode.lineno < node.lineno)
+                )
+            ):
+                # Relation of a name to the same name in a named expression
+                # Could be used before assignment if self-referencing:
+                # (b := b)
+                # Otherwise, safe if used after assignment:
+                # (b := 2) and b
+                maybe_before_assign = defnode.value is node or any(
+                    anc is defnode.value for anc in node.node_ancestors()
+                )
+
+    return maybe_before_assign, annotation_return, use_outer_definition
+
+
+def _maybe_used_and_assigned_at_once(defstmt: _base_nodes.Statement) -> bool:
+    """Check if `defstmt` has the potential to use and assign a name in the
+    same statement.
+    """
+    if isinstance(defstmt, nodes.Match):
+        return any(case.guard for case in defstmt.cases)
+    if isinstance(defstmt, nodes.IfExp):
+        return True
+    if isinstance(defstmt, nodes.TypeAlias):
+        return True
+    if isinstance(defstmt.value, nodes.BaseContainer):
+        return any(
+            _maybe_used_and_assigned_at_once(elt)
+            for elt in defstmt.value.elts
+            if isinstance(elt, (*NODES_WITH_VALUE_ATTR, nodes.IfExp, nodes.Match))
+        )
+    value = defstmt.value
+    if isinstance(value, nodes.IfExp):
+        return True
+    if isinstance(value, nodes.Lambda) and isinstance(value.body, nodes.IfExp):
+        return True
+    if isinstance(value, nodes.Dict) and any(
+        isinstance(item[0], nodes.IfExp) or isinstance(item[1], nodes.IfExp)
+        for item in value.items
+    ):
+        return True
+    if not isinstance(value, nodes.Call):
+        return False
+    return any(
+        any(isinstance(kwarg.value, nodes.IfExp) for kwarg in call.keywords)
+        or any(isinstance(arg, nodes.IfExp) for arg in call.args)
+        or (
+            isinstance(call.func, nodes.Attribute)
+            and isinstance(call.func.expr, nodes.IfExp)
+        )
+        for call in value.nodes_of_class(klass=nodes.Call)
+    )
+
+
+def _is_only_type_assignment(
```

Comment:
If we go through with this, let's add this commit to the git ignore revs list, and we'll be sure to merge it with a merge commit to avoid rewriting the hash.